### PR TITLE
BLOCKED/WIP: Add table for LTP results

### PR DESCRIPTION
**Pull request by Martchus**
WIP: Add table for LTP results to test details page

* Tab 'LTP' is only available when LTP results present
  and only loaded on demand
* Is basically just queries the JSON and renders the table
  on the client using JavaScript/DataTables
* See https://progress.opensuse.org/issues/25680

---

still TODO:
* Improve filtering
* Map LTP results to OpenQA result strings
* Count runs / fails when test is executed multiple times
  (similar to http://atrey.karlin.mff.cuni.cz/~metan/outgoing/ltp-syscalls.html), **or is this not required here**?
* Log errors (and show flash in error case)
* Add test
* Use a less generic name for the JSON file (must be provided by LTP tests of
  course)

---

screenshot_20171107_173134
screenshot_20171107_173205

### CODE REVIEW

**Comment by coolo** [comment]
I don't like this specific LTP support at all. If we can't map this to something generic, it's not going to happen.

I don't mind a javascript solution to render test generated data, but it needs to be a documented format and it needs to be generated by the tests preparing for rendering. I bet there is some way to map this into junit format.

**Comment by Martchus** [comment]
> it needs to be a documented format

That is true and the lack of documentation is also what prevents myself from further developing this. (Eg. questions like what are possible results and what do they mean came up.)

> I bet there is some way to map this into junit format.

I don't know junit format, but mapping to something generic sounds like a good idea.

---

> it needs to be generated by the tests preparing for rendering

I agree - if we would implement this mapping inside the openQA web application we would gain nothing from it. So I suppose this PR/issue is blocked until the tests generate the results in a more generic way.

**Comment by schlad** [comment]
"So I suppose this PR/issue is blocked until the tests generate the results in a more generic way." - this should be easy to resolve.
@coolo what do we need to provide? Current solution prepares json. Is that good enough provided we have this documented?

@coolo and it seems to me like we are really aiming at solution discussed here between Marius and Metan: https://progress.opensuse.org/issues/25680.
"Option (4) would be adding an openQA API to report structured test results for a console based tests and build a result table"
"It would be a different table (in another tab) as well but also adds API functions specialized for console tests."

**Comment by coolo** [comment]
I don't think we need to invent a new test result format - there are already a million of it. I don't mind if we had a less screenshot like view for it, but I do mind us guessing about possible LTP glitches.

So pick a standard format (I take xunit as https://lukejpreston.github.io/xunit-viewer/ gives you an idea how it could be rendered for LTP), make LTP tests generate it and then we talk about simplified rendering of it.

**Comment by metan-ucw** [comment]
@coolo let me explain the situation a bit more.

First of all the current result format, as presented by openQA, is actually actively harming the review process of console based testcases and hence the SLE15 development. It's worse than suboptimal, it's painful to use. So maybe it makes sense to merge suboptimal solution for now as that would help us with review of at least half of the test runs we have in openQA. Let me use an analogy here. Imagine that you have leaky toilet pipe, the replacement is ordered but it will take month to arrive. The logical action is to fix the issue temporarily by taping the cracks with a duct tape, but you are telling me to wait for the proper replacement and to mop down the s**t from the floor every day.

Secondly LTP is not the only one, it was only used as foster child, since it was one of the first large testsuites we integrated into the openQA. But we really need a proper way for integrating large command line testsuites and it has to be core openQA feature. If we are supposed to use openQA for our daily testing we need to be treated as first-class citizens there as well.

**Comment by coolo** [comment]
you are treated as first-class citizens. I'm just expecting you to sit on the toilet while shitting - that's all.

**Comment by richiejp** [comment]
I have documented this, but it is not finished so it will change. Please see the following:

https://progress.opensuse.org/issues/17048
https://gitlab.suse.de/rpalethorpe/srd

The reason I am using JSON is because it is the most commonly used format in big data analyses, which makes it very easy to use with tools like MongoDB and Apache Spark. It allows for a semi-rigid schema which allows for structured test specific data to be stored and analysed easily. However  I would also be interested in any standardised-self-describing formats (i.e. it has meta data to describe the data's structure and is widely used) so long as they are not needlessly verbose or complex.

So, the thing you are having a problem with, is actually a feature. I have always know this would be a problem to implement in OpenQA because it lacks the infrastructure to handle flexible data which is why I am exporting the data and analysing it in MongoDB using a bunch of scripts.

However, all is not lost, because we can also export to a simpler rigid format for the sake of software like OpenQA... which is most software.

@Martchus What format do you want? Maybe XUNIT will work, but it makes my eyes bleed. Also we will probably still need to define which elements are required because it appears to be flexible to some extent.

Also please note that this is something I have been doing during the kernel review and have been attempting to keep the time I am spending on it to a minimum. I am only doing it because I really need a flexible data storage and analyses solution to do the LTP kernel review efficiently. I have created a solution which works for me, but it is basically an experiment.

Also, as an extra complication, it would be really great if we could comment on (and tag) individual LTP test case failures. Then doing the LTP review in OpenQA with Oliver's script would become viable. Which is useful if someone other than me needs to do the review.

**Comment by Martchus** [comment]
> What format do you want?

To answer that question, I would have to look into different options myself as well. I can do that next time at work. Then I'll also checkout the links.

In general I'm not very picky, however the following to points would ease my work:
* The format is well documented regarding its structure, meaning of the results, ...
* It is JSON, simply because than I can stick with the current approach and render in JavaScript.
* Of course it should be capable of holding all the information *you* expect to be present in openQA (to prevent the need of having extra solutions after all).
* > so long as they are not needlessly verbose or complex

Note that maybe @coolo has a stronger opinion on the concrete format to be used.

---

> Also, as an extra complication, it would be really great if we could comment on (and tag) individual LTP test case failures.

This would be a lot of effort but doable. But if you think that far, we really need to rely on a standardized format. Building such a feature on top of some very specific extensions is not a good idea.

**Comment by okurz** [comment]
> Also, as an extra complication, it would be really great if we could comment on (and tag) individual LTP test case failures. Then doing the LTP review in OpenQA with Oliver's script would become viable. Which is useful if someone other than me needs to do the review.

The intention of [okurz/openqa_review](http://github.com/okurz/openqa_review) was beside some other things that it can act as a "rapid prototyping platform" for openQA review features. So maybe it *is* easier to hack something into that script to extract the individual LTP test case failure but your "srd" approach might already do the same thing already just as well.

**Comment by coolo** [comment]
If we want to have tag overtake, we need to have openQA a clear picture on what failed and a javascript only solution will not cover it.

I'm basically objecting a middle ground. What is proposed here is that most stuff is handled in an external tool and we inject some specific hack into openQA to make it easier to have the output required by this tool displayed.

Either we do the review details on that external tool (fine with me, as long as the results of the review are written back into the dashboard comments - manually or automatically) or we do it properly within openQA by defining a format tests upload (we already have junit for slenkins tests and others, but we can extend it with another format too). But the later requires more work than a JS hack - and practically also more than what we can commit to atm. But it's a goal I commit to in general.

**Comment by richiejp** [comment]
Well to start with, I should probably fork your script and convert the srd stuff into Python. There is some functionality in your script that I can reuse and Python is probably the better choice for this. Also you may be able to adapt what I have done for regular OpenQA tests.

Then it shouldn't be hard to change your script to parse, say "sssd(blame=ltp): bsc#810395" where 'sssd' is an expression describing one or more LTP* test names and '(blame=ltp)' is some meta data. So it will tag the specific LTP test failure with that bug instead of the OpenQA test module. Otherwise, in order to use my MongoDB report you have to write some javascript which adds the tag, like `ref_bug('bsc', '1040381', /LTP:mm:oom.*/, { blame: 'kernel' });`.

It would still be nicer to be able to comment on individual LTP test results, but that can wait.

* This can also be applied to other test suites, such as the XFS tests. They just need to export the results in the same way as the LTP.

**Comment by richiejp** [comment]
>If we want to have tag overtake, we need to have openQA a clear picture on what failed and a >javascript only solution will not cover it.

Don't worry about tag overtake.

>I'm basically objecting a middle ground. What is proposed here is that most stuff is handled in an >external tool and we inject some specific hack into openQA to make it easier to have the output >required by this tool displayed.

We are avoiding making a solution specific to LTP.

>Either we do the review details on that external tool (fine with me, as long as the results of the review >are written back into the dashboard comments - manually or automatically) or we do it properly within >openQA by defining a format tests upload (we already have junit for slenkins tests and others, but we >can extend it with another format too). But the later requires more work than a JS hack - and >practically also more than what we can commit to atm. But it's a goal I commit to in general.

From the screenshots the JS 'hack' solves the immediate problem of just displaying the results in a way people can understand for minimal development cost and is run client side in a format native to the browser. The Slenkins tests are also unreadable, so they can also benefit from this.

**Comment by coolo** [comment]
For what it's worth: I agree with sleninks tests being unreadable.

**Comment by metan-ucw** [comment]
I agree with @richiejp the actual hack solves the immediate problem quite nicely, what @coolo describes is a proper long term solution which should eventually replace it. But as I tried to explain already we need to be able to display the results now, actually we needed it ready half a year ago. In the end people look at the results and do the review, so for me having a usable interface between openQA and actual human is a priority.

Hence I'm all for choosing well defined JSON based format and having it displayed with Javascript for now.

**Comment by richiejp** [comment]
@Martchus I am wondering if you would like the top-level status field to be changed to a ternary value i.e pass, fail and skipped. Otherwise you will need to read LTP specific details to decide if a test failed with conf. It seems quite common to have three values.

Also I am thinking of making "log" a top level field (although maybe an optional one) as I think most tests will have some kind of raw output that can be displayed with no parsing. Again, it seems quite common amongst test formats.

**Comment by okurz** [comment]
Talked with @foursixnine @schlad @pevik @richiejp:
1. openQA should not try to do the job of specialized dashboards for evaluation, e.g. fancy graphs, etc.
2. xunit (not the .NET framework) can be used. Would be nice if os-autoinst would provide a testapi function
3. the UX for displaying individual test step results can be improved when it's not about "just screenshots". That is what this PR is about

**Comment by richiejp** [comment]
On point 2. I was thinking of having a function like:

`external_result(test_name, pass | fail | skipped, log [, arbitrary_data ])`

In os-autoinst's testapi.pm. This will then be submitted when the test finishes to OpenQA and displayed in an extra tab, similar to the screenshot above.

Note that the function name is just for illustration purposes.

**Comment by Martchus** [comment]
> the UX for displaying individual **test step results** can be improved when it's not about "just screenshots". That is what this PR is about

Displaying the individual steps for "non-screenshot tests" could be improved. However, I'm not sure how this is supposed to work in a general way. To get something like the table proposed by this PR some LTP specific logic would be required as well. For example we would need to figure which details belong together. This seems to be more effort that just making the table directly from the JSON file. Hence the decision for some JSON-based format.

*So this PR is about displaying further test details provided as some kind of JSON* (which should be some standardized, documented format).

> If we want to have tag overtake, we need to have openQA a clear picture on what failed and a
> javascript only solution will not cover it.

It wasn't clear to me that this would be a requirement. Of course than we would need process the information on the server-side.

---

@richiejp So you're experimenting with a openQA-review-like solution? In fact you just would need to download the JSON from openQA and render a table similar like the PR does. A simple, single-page application which does some AJAX would do. You could even display/alter the openQA comments there to do some tagging.

Hence, when talking with @metan-ucw I asked the question:

> What is the benefit of having this *inside* openQA? Why can't you just use an external tool/website?

Maybe @metan-ucw can explain again what the benefit of having this inside openQA would be?

---

@richiejp Implementing an API like you proposed in the last comment could also be a solution. Then the test writer would not have to care which format is *internally* used by openQA to store the data.

What should openQA do with that `arbitrary_data`? Providing a download link? Maybe a mime-type for that data would not hurt. The [initial reference table](http://atrey.karlin.mff.cuni.cz/~metan/outgoing/ltp-syscalls.html) also includes runtime. Would that be required, too? Then the API function could look like:

```
external_result(test_name, pass | fail | skipped, log [, started, finished , arbitrary_data, arbitrary_data_mime_type])
```

---

To summarize: So far we gathered a lot of ideas, however it is not clear to me what the requirements are now and depending on this whether we should stick to the current approach (client-side JSON rendering).

Alternatives:
* server-side processing of JSON
* a new test-API function (which *might* use internally a standardized format as well)

However, it should be clear that if we stick to JSON, we need to decide which particular format it should be. Proposed formats:
* junit
* xunit

**Comment by richiejp** [comment]
> It wasn't clear to me that this would be a requirement. Of course than we would need process the information on the server-side.

It is not a requirement at this time, but it may be in the future. So you will probably want to consider it in your design. However for now we can handle it externally.

> 
> So you're experimenting with a openQA-review-like solution? In fact you just would need to download the JSON from openQA and render a table similar like the PR does. A simple, single-page application which does some AJAX would do. You could even display/alter the openQA comments there to do some tagging.
> 
> Hence, when talking with @metan-ucw I asked the question:
> 
>     What is the benefit of having this inside openQA? Why can't you just use an external tool/website?
> 
> Maybe @metan-ucw can explain again what the benefit of having this inside openQA would be?

I have a bunch of scripts which import data into MongoDB and run some complex queries to produce a report. For reasons I will skip over, I don't think there is a way to make this accessible to everyone without creating a full web application/service which queries and updates the data.

I think that creating such an application is the best thing to do, but not everyone agrees with me and right now we just need the basic results to be displayed in a sane way, where everyone can access them easily. Everyone at least knows where and what OpenQA is and it is the default application for test automation.

> Implementing an API like you proposed in the last comment could also be a solution. Then the test writer would not have to care which format is internally used by openQA to store the data.
> 
> What should openQA do with that arbitrary_data? Providing a download link? Maybe a mime-type for that data would not hurt. The initial reference table also includes runtime. Would that be required, too? Then the API function could look like:

It is deliberately left undefined, it could be anything so long as it is represented in such a way that it can be exported by OpenQA in various different formats. To begin with you can ignore it and then later on use it to expand the API if necessary.

If you would like _started_ and _finished_ to be part of the initial specification, then they can be defined as optional fields like

`external_result( name, ..., [ started => <time>, finished => <time>, other => <HashMap> ])` 

What data a test can produce is open ended, so the API for submitting data should be open ended as well. I guess if you allow raw byte arrays to be included in the data then it would be useful to provide an optional MIME/type _hint_ with them. Again that is not an immediate concern though. As long as the API has an extension point we can add it later.

> However, it should be clear that if we stick to JSON, we need to decide which particular format it should be. Proposed formats:
> 
>     junit
>     xunit

You can make the JSON a superset of both those formats. As far as I can see, you will need to convert the results to some format to send them from os-autoinst to OpenQA and then from OpenQA to the browser. Os-autoinst already communicates with OpenQA using JSON and browsers natively support JSON, so it might as well be JSON the whole way. You just need to make sure the JSON contains all the necessary information to be converted to Xunit and that XUnit can be converted to the JSON.

However if I am just passing a Perl object to an API in os-autoinst then I don't care[1] what format is then used internally so long as it does not restrict the API. So now I think the best thing to do is prototype and evaluate the API.

[1] At least in theory.

**Comment by mudler** [comment]
I do think that the approach proposed by @coolo is the more clean, and i don't see why it should be blocking - what we need here is just to be able to upload 'extra tests' results and store them separately in the current one (while keeping the relationship in the database) in a common format ( as xunit seems to be the ideal one ).  

Why JSON? that would require us to re-invent the wheel again and decide what works for everybody (and will likely end up as a port of xunit in JSON) over and over. 

If the problem is how to store the xunit on the openQA side, there are plenty options: serialization into DB (as it is xml, makes sense) , upload as a file asset, ... you name it .

The WebUI then can parse xunit results and render it as plain HTML (if present in the database) and add the 'extra tests' results tab on need -xunit results than can point even further to other tests assets that can be uploaded as part as the test as well.

**Comment by Martchus** [comment]
Ok, so let it be xunit. XML is ok, too. It is also easy to parse it client-side but maybe we want server-side parsing after all? I mean, if this becomes a 'big, generic feature', it might be worth it.

So the first step would be providing some test API "to be able to upload 'extra tests' results". Whether we want to store the xunit in the database (which means server-side handling of the xunit data) or just as it is on the file system depends on whether openQA should to any further processing of the data (comments, tagging, etc.). Since ideas like that came up, I have to ask whether this is a requirement or not?  
(Note that simple filering/sorting can easily be done on the client-side.)

**Comment by coolo** [comment]
I actually believe that for a proper review process, you need proper failure carry over. And for this you need to differentiate the test failing. But in the context of console tests, the test failing is defined in xunit and not in the openQA test.

So I do believe we need:
1. stop supporting the parsing of junit in isotovideo
2. either add an explicit API for test results or rely on a filename pattern within uploaded logs, but provide means to add xunit to the test results
3. have the webapi detect failed xunit tests and mark them as test modules (in fact the passing ones are boring :)
4. render the data more up to the task. the client side could actually be JSON based if we have something serializing the xunit into the DB anyway, we can just as well pull it into a JSON blob within PG (https://www.postgresql.org/docs/9.4/static/functions-json.html)

**Comment by coolo** [comment]
http://rpws.suse.cz/tests/134 splits the tests on loadtest. So is there actually any need for one big xunit in that scenario? @richiejp

**Comment by mudler** [comment]
Indeed, i never said that proper failure carry over was gone ;) - this can be granted after webui parsed the xunit received by the worker - so the extra tests results would have a mapping to an actual openQA test result.

**Comment by coolo** [comment]
what I'm not sure about is though if /tests/134 is actually the optimum or rather a workaround.

**Comment by foursixnine** [comment]
@coolo I think it looks more like a workaround... http://rpws.suse.cz//api/v1/jobs/134/details will give you the json data with the details/results and stuff, and you can navigate to /job/testresults/ to look for the information which solves the need for the json data export~

**Comment by richiejp** [comment]
@coolo It is much better, especially if the results were 'collapsed' by default, but it is not quite good enough. We probably at least want to be able to filter based on test name and status and choose which column to sort by.

Failure carry over doesn't work with the LTP under the current system. To get it to work properly we need to be able to specify the kind of environment the bug is likely to happen in as well as which tests are likely to cause it. So basically provide a key-value set describing the environment of the failure and the tests which failed and compare that with another key-value set describing the bug. Possibly in some cases a more advanced comparison would be necessary, so the user would need some way of describing more advanced logic. We would definitely like this (it is what my MongoDB scripts are trying to do), but we don't want it to block having the basic table either.

As far as I can tell XUnit does not allow for top level environment information, so it is just not suitable at all. JUnit does although it doesn't allow for a 1:1 conversion to my format because my format allows arbitrary structure within the environment section whereas in JUnit this appears to be restricted to a flat key-value set in the testsuite section. Furthermore it doesn't appear to support per test case attributes/properties which are not already defined in the schema. In my format you can include arbitrary-structured-test-specific data which is necessary for more advanced analysis and can be ignored the rest of the time.

I am not just reinventing the wheel because I can, there are actual reasons for it. Again, see https://progress.opensuse.org/issues/17048 (it is a bit out of date now, but might help). However what is more important right now is that we get the table display. I can continue to export my own format regardless.

**Comment by mudler** [comment]
so, to ease this out -  i'm working in a generic parsing solution that allows at least to include external test in an (almost) uniform way, translating them in an internal test representation that can be serialized and deserialized -  with the possibility to extend it with different formats and also to map them as openQA test modules run if needed.

That's my branch https://github.com/mudler/openQA/tree/test_parser , currently the parser is able to read/write/translate JUnit - will extend it to the JSON format proposed by @richiejp and XUnit - so we can use it to serialize/deserialize external tests results inside the openQA DB.

**Comment by Martchus** [comment]
So I suppose I'll make use of https://github.com/os-autoinst/openQA/pull/1545 when it is ready.

**Comment by Martchus** [comment]
Since #1545 has been merged, I'm currently checking how I can use it for this task. My findings so far:

Seems like you have already been created a parser for LTP results. So I could use `lib/OpenQA/Parser/Format/LTP.pm` for parsing `result_array.json`. However, it seems like you have also already implemented triggering the parser on upload (`sub parse_extra_tests`).

So the first steps for a further PR implementing https://progress.opensuse.org/issues/25680 would be:
* [ ] Provide a method in `Jobs.pm` to conveniently access the data written via `$parser->write_output($self->result_dir);` from controller/template code.
* [ ] Create a chrome driver test case where a job with LTP results is present.
* [ ] If results are available, create a new tab
* [ ] and display the information similar like the current PR does.
    * Provide a route which returns the data for the datatable as JSON (but this time based on the generic data structure provided by #1545).
    * *Or* render the data directly as HTMl using a template.

---

Outside of the scope of this PR:
* [ ] Make the LTP tests actually upload the results in the required way.

---

@mudler Does this make sense or have I overlooked something your PR already provides?

**Comment by mudler** [comment]
Makes all sense, just few points:

- you can also just use the parser module and do not need to load the specific implementation when using it. This is especially useful when you don't know what type of file you are going to parse, and avoids the need to import all specific formats supported.

```perl
use OpenQA::Parser 'parser';

my $data = parser(LTP => 'result_array.json'); 

# $data is a OpenQA::Parser::Format::LTP object now
# usage is shown in POD, but ->results() returns a Mojo::Collection
# all methods from it are valid: $data->results->each , $data->results->first etc.. (just watch out it's a collection of objects)

# So you can:

my $type = 'LTP';
my $data = parser($type => 'file'); 

# and avoid to import the single packages at all
use OpenQA::Parser::Format::LTP;

```

- The API endpoint implemented in that PR will update the test running modules with the extra test file, but it's not saving the parsed data (in json/storable) from the parser nor saving the type of the file that was parsed. 
The behavior enabled by the PR is triggered only if in the api call the parameter ```extra_test``` is set to ```1```. 
 e.g. :  
```perl
$t->post_ok(
        '/api/v1/jobs/99963/artefact' => form => {
            file       => {file => $file, filename => $fname},
            type       => "LTP",
            extra_test => 1,
            script     => 'test'
        })->status_is(200);
```

That means, what needs still to be implemented is also the database table that will hold the serialized extra tests sent by a worker - if we want the WebUI to freely 'browse' generic extra results.
Or, the file can be also parsed again instead of storing the serialization, and the client call a webui endpoint that sends back the re-constructed json.

**Comment by Martchus** [comment]
> Or, the file can be also parsed again instead of storing the serialization, and the client call a webui endpoint that sends back the re-constructed json.

My idea was to store the generic serialization on the file system. Like the initial PR but this time using a generic format your parser can export. This file could be directly delivered to the JavaScript on the client. There is no additional server-load for re-constructing JSON on the server.

---

```
        $parser->load($tmp_extra_test)->results->each(
            sub {
                return if !$_->test;
                $_->test->script($script) if $script;
                my $t_info = $_->test->to_openqa;
                $self->insert_module($t_info);
                $self->update_module($_->test->name, $_->to_openqa);
            });
```

Seems like it actually adds the parsed results as job modules. Maybe I can also just show these modules in a different way if there's no required information lost.

---

> but it's not saving the parsed data

I'm wondering what this line is supposed to do then:
```
        $parser->write_output($self->result_dir);
    };
```



---

> That means, what needs still to be implemented is also the database table that will hold the serialized extra tests sent by a worker

We can also extend the database. It would have the advantage that we could attach further information to it and have efficient server-side filtering/sorting.  
However, I don't think that's required. I suppose adding the parsed data as job modules already covers the 'failure carry over' use case as the failure is recorded as failing job module.

**Comment by mudler** [comment]
fine by me, i'm just proposing on how to use it ;)

```
$parser->write_output($self->result_dir);
```

It's a convenience function to extract the tests output (as real console logs) and write them in the specified directory, as the worker would do it, see https://github.com/mudler/os-autoinst/commit/c637fc1cafba9db3d690f30a305eb15567542a40 as an example (the change it's using the parser from the worker, instead that from the webui) yielding to this results: http://e122.suse.de/tests/5542

**Comment by Martchus** [comment]
We merged already https://github.com/os-autoinst/openQA/pull/1639 and adding a separate table is still conceivable. However, I don't think it makes sense to base anything on this PR anymore, so I'm closing.
